Kernel panic: "Debug cannot be handled via signals" when Trap Flag (TF) is set in user-space RFLAGS

### Describe the bug

A kernel panic occurs when a user-space program sets the Trap Flag in the RFLAGS register and executes a syscall instruction.

The panic is triggered because the Asterinas kernel's exception handling logic does not currently support mapping the Debug Exception to a user-space signal (such as SIGTRAP). Instead, the kernel invokes panic! when it encounters a Debug exception that it cannot handle.

# https://github.com/asterinas/asterinas/blob/b7ae7383102800529a079f00f92178511d35b8f3/kernel/src/arch/x86/signal.rs#L40

### To Reproduce

1. Compile the program and run.

```
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/syscall.h>
#include <errno.h>

int main() {
    printf("Attempting to trigger kernel panic through syscall mechanism...\n");
    
    // Try to trigger the RFlags issue mentioned in the original description
    // Note: This is user space, so won't directly trigger kernel init panic
    // but attempts to exercise the syscall path
    asm volatile (
        "push $0x47700\n\t"
        "popf\n\t"
        "syscall\n\t"
    );

    printf("If kernel is vulnerable, panic may occur during boot initialization\n");
    printf("Note: This PoC cannot directly trigger the init_on_cpu() panic from user space\n");
    
    return 0;
}
```

### Environment

- Official docker asterinas/asterinas:0.17.0-20251217
- Asterinas version: main [[Cleanup: Only maintain children for directory dentries · asterinas/asterinas@b7ae738](https://github.com/asterinas/asterinas/commit/b7ae7383102800529a079f00f92178511d35b8f3)](https://github.com/asterinas/asterinas/commit/b7ae7383102800529a079f00f92178511d35b8f3)

### Logs

```
~ # ./ex611___home__workspace__asterinas__ostd__src__arch__x86__trap__syscall.rs_59_23 
Attempting to trigger kernel panic through syscall mechanism...
[    55.876] ERROR: Uncaught panic:
        Debug cannot be handled via signals (Debug)
        at /root/asterinas/kernel/src/arch/x86/signal.rs:40
        on CPU 0 by thread Some(Thread { task: (Weak), data: Any { .. }, is_exited: false, cpu_affinity: AtomicIdSet { bits: [1], phantom: PhantomData<ostd::cpu::id::CpuId> }, sched_attr: SchedAttr { policy: SchedPolicyState { kind: AtomicSchedPolicyKind(2), policy: UnsafeCell { .. } }, last_cpu: AtomicCpuId(0), real_time: RealTimeAttr { prio: 99, time_slice: 31490340 }, fair: FairAttr { weight: 1024, pending_weight: 0, vruntime: 897488727 } } })
Printing stack trace:
   1: fn 0xffffffff889dfad0 - pc 0xffffffff889dfb1b / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd83fd0;

   2: fn 0xffffffff8804b7c0 - pc 0xffffffff8804bfbd / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd83ff0;

   3: fn 0xffffffff8804e600 - pc 0xffffffff8804e60a / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd843c0;

   4: fn 0xffffffff8804b000 - pc 0xffffffff8804b00a / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd843d0;

   5: fn 0xffffffff88aec900 - pc 0xffffffff88aec946 / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd843e0;

   6: fn 0xffffffff884bf7b0 - pc 0xffffffff884bf881 / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84420;

   7: fn 0xffffffff88591900 - pc 0xffffffff8859191d / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84500;

   8: fn 0xffffffff88591580 - pc 0xffffffff885916e3 / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84550;

   9: fn 0xffffffff88240000 - pc 0xffffffff8824096a / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84650;

  10: fn 0xffffffff88240a70 - pc 0xffffffff88240ace / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84b40;

  11: fn 0xffffffff885bf9f0 - pc 0xffffffff885bfa1e / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84cf0;

  12: fn 0xffffffff885bf920 - pc 0xffffffff885bf966 / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84d20;

  13: fn 0xffffffff88276d10 - pc 0xffffffff88276d3c / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84d90;

  14: fn 0xffffffff88685370 - pc 0xffffffff886853af / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84dd0;

  15: fn 0xffffffff88240a30 - pc 0xffffffff88240a57 / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84f40;

  16: fn 0xffffffff88320f10 - pc 0xffffffff88320f2d / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84f80;

  17: fn 0xffffffff889db7f0 - pc 0xffffffff889db815 / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84fa0;

  18: fn 0xffffffff889e07e0 - pc 0xffffffff889e0864 / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd84fd0;

  19: fn 0xffffffff889fb4d1 - pc 0xffffffff889fb4da / registers:

     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd85000;


     rax 0xffffdfffffd83ee8; rdx 0xffffffff889f8402; rcx  0x1000002010024fa; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdbfd0; rbp                0x0; rsp 0xffffdfffffd85008;

OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) ostd::panic::print_stack_trace
          at /root/asterinas/ostd/src/panic.rs:113
(  2) aster_kernel::thread::oops::panic_handler
          at /root/asterinas/kernel/src/thread/oops.rs:130
(  3) __ostd_panic_handler
          at ??:?
(  4) __rustc::rust_begin_unwind
          at blp0687gyqpr7q22lmo8m09ck:?
(  5) core::panicking::panic_fmt
          at ??:?
(  6) <aster_kernel::process::signal::signals::fault::FaultSignal as core::convert::From<&ostd::arch::cpu::context::CpuException>>::from
          at /root/asterinas/kernel/src/arch/x86/signal.rs:20
(  7) aster_kernel::thread::exception::generate_fault_signal
          at bibuuqa0kib35pv0g3wzkyuwf:?
(  8) aster_kernel::thread::exception::handle_exception
          at /root/asterinas/kernel/src/thread/exception.rs:45
(  9) aster_kernel::thread::task::create_new_user_task::{closure#0}
          at 7bt8xmraeupn8esabxu01zr0k:?
( 10) aster_kernel::thread::task::create_new_user_task::{closure#1}
          at /root/asterinas/kernel/src/thread/task.rs:126
( 11) unwinding::panicking::catch_unwind::do_call::<aster_kernel::thread::task::create_new_user_task::{closure#1}, ()>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panicking.rs:56
( 12) unwinding::panicking::catch_unwind::<unwinding::panic::RustPanic, (), aster_kernel::thread::task::create_new_user_task::{closure#1}>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panicking.rs:42
( 13) unwinding::panic::catch_unwind::<(), aster_kernel::thread::task::create_new_user_task::{closure#1}>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panic.rs:87
( 14) aster_kernel::thread::oops::catch_panics_as_oops::<aster_kernel::thread::task::create_new_user_task::{closure#1}, ()>
          at /root/asterinas/kernel/src/thread/oops.rs:55
( 15) aster_kernel::thread::task::create_new_user_task::{closure#2}
          at /root/asterinas/kernel/src/thread/task.rs:131
( 16) <aster_kernel::thread::task::create_new_user_task::{closure#2} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250
( 17) <alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output = ()> + core::marker::Send> as core::ops::function::FnOnce<()>>::call_once
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2206
( 18) kernel_task_entry
          at /root/asterinas/ostd/src/task/mod.rs:196
( 19) kernel_task_entry_wrapper
          at ??:?
make: *** [Makefile:308: run_kernel] Error 1
```

--- a/kernel/src/arch/loongarch/signal.rs
+++ b/kernel/src/arch/loongarch/signal.rs
@@ -1,8 +1,11 @@
 // SPDX-License-Identifier: MPL-2.0
 
+use loongArch64::register::estat::Exception;
 use ostd::arch::cpu::context::{CpuExceptionInfo, UserContext};
 
-use crate::process::signal::{SignalContext, sig_num::SigNum, signals::fault::FaultSignal};
+use crate::process::signal::{
+    SignalContext, constants::*, sig_num::SigNum, signals::fault::FaultSignal,
+};
 
 impl SignalContext for UserContext {
     fn set_arguments(&mut self, sig_num: SigNum, siginfo_addr: usize, ucontext_addr: usize) {
@@ -13,7 +16,30 @@ impl SignalContext for UserContext {
 }
 
 impl From<&CpuExceptionInfo> for FaultSignal {
-    fn from(_trap_info: &CpuExceptionInfo) -> Self {
-        unimplemented!()
+    fn from(trap_info: &CpuExceptionInfo) -> Self {
+        let (num, code, addr) = match trap_info.code {
+            Exception::LoadPageFault | Exception::StorePageFault | Exception::FetchPageFault => {
+                (SIGSEGV, SEGV_MAPERR, Some(trap_info.page_fault_addr as u64))
+            }
+            Exception::PageModifyFault
+            | Exception::PageNonReadableFault
+            | Exception::PageNonExecutableFault
+            | Exception::PagePrivilegeIllegal => {
+                (SIGSEGV, SEGV_ACCERR, Some(trap_info.page_fault_addr as u64))
+            }
+            Exception::FetchInstructionAddressError | Exception::MemoryAccessAddressError => {
+                (SIGBUS, BUS_ADRERR, None)
+            }
+            Exception::AddressNotAligned => (SIGBUS, BUS_ADRALN, None),
+            Exception::BoundsCheckFault => (SIGSEGV, SEGV_BNDERR, None),
+            Exception::Breakpoint => (SIGTRAP, TRAP_BRKPT, None),
+            Exception::InstructionNotExist | Exception::InstructionPrivilegeIllegal => {
+                (SIGILL, ILL_ILLOPC, None)
+            }
+            Exception::FloatingPointUnavailable => (SIGFPE, FPE_FLTINV, None),
+            Exception::Syscall | Exception::TLBRFill => unreachable!(),
+        };
+
+        FaultSignal::new(num, code, addr)
     }
 }

--- a/kernel/src/arch/x86/signal.rs
+++ b/kernel/src/arch/x86/signal.rs
@@ -23,6 +23,8 @@ impl From<&CpuException> for FaultSignal {
             }
             CpuException::BoundRangeExceeded => (SIGSEGV, SEGV_BNDERR, None),
             CpuException::AlignmentCheck => (SIGBUS, BUS_ADRALN, None),
+            CpuException::Debug => (SIGTRAP, TRAP_TRACE, None),
+            CpuException::BreakPoint => (SIGTRAP, SI_KERNEL, None),
             CpuException::InvalidOpcode => (SIGILL, ILL_ILLOPC, None),
             CpuException::GeneralProtectionFault(..) => (SIGBUS, BUS_ADRERR, None),
             CpuException::PageFault(raw_page_fault_info) => {

--- a/kernel/src/process/signal/signals/fault.rs
+++ b/kernel/src/process/signal/signals/fault.rs
@@ -1,7 +1,5 @@
 // SPDX-License-Identifier: MPL-2.0
 
-#![cfg_attr(target_arch = "loongarch64", expect(dead_code))]
-
 use super::Signal;
 use crate::{
     prelude::*,

--- a/test/initramfs/src/regression/process/run_test.sh
+++ b/test/initramfs/src/regression/process/run_test.sh
@@ -51,6 +51,7 @@ set -e
 
 if [ "$(uname -m)" = "x86_64" ]; then
     ./signal/sigaltstack
+    ./signal/sigtrap
     ./signal/signal_fpu
     ./signal/signal_rflags_df
     ./signal/signal_test

--- a/test/initramfs/src/regression/process/signal/Makefile
+++ b/test/initramfs/src/regression/process/signal/Makefile
@@ -3,6 +3,7 @@
 ifneq ($(HOST_PLATFORM), x86_64-linux)
 C_OBJS_FILTER += \
 	sigaltstack \
+	sigtrap \
 	signal_fpu \
 	signal_rflags_df \
 	signal_test

--- a/test/initramfs/src/regression/process/signal/sigtrap.c
+++ b/test/initramfs/src/regression/process/signal/sigtrap.c
@@ -1,0 +1,67 @@
+// SPDX-License-Identifier: MPL-2.0
+
+// Regression test for mapping #DB and #BP exceptions to SIGTRAP.
+
+#define _GNU_SOURCE
+#include <signal.h>
+#include <string.h>
+#include <ucontext.h>
+#include <unistd.h>
+
+#include "../../common/test.h"
+
+static volatile int trap_brkpt_received;
+static volatile int trap_trace_received;
+
+static void sigtrap_handler(int sig, siginfo_t *info, void *ctx)
+{
+	if (sig != SIGTRAP) {
+		fprintf(stderr, "expected SIGTRAP, got %d\n", sig);
+		_exit(1);
+	}
+
+	if (info->si_code == SI_KERNEL) {
+		trap_brkpt_received = 1;
+	} else if (info->si_code == TRAP_TRACE) {
+		ucontext_t *uc = (ucontext_t *)ctx;
+		uc->uc_mcontext.gregs[REG_EFL] &= ~0x100UL;
+		trap_trace_received = 1;
+	}
+}
+
+static int trap_brkpt(void)
+{
+	trap_brkpt_received = 0;
+	asm volatile("int3");
+	return 0;
+}
+
+static int trap_trace(void)
+{
+	trap_trace_received = 0;
+	asm volatile("pushfq; orq $0x100,(%%rsp); popfq; nop"
+		     :
+		     :
+		     : "memory", "cc");
+	return 0;
+}
+
+FN_SETUP(init)
+{
+	struct sigaction sa;
+
+	memset(&sa, 0, sizeof(sa));
+	sa.sa_sigaction = sigtrap_handler;
+	sa.sa_flags = SA_SIGINFO;
+	sigemptyset(&sa.sa_mask);
+
+	CHECK(sigaction(SIGTRAP, &sa, NULL));
+}
+END_SETUP()
+
+FN_TEST(sigtrap)
+{
+	TEST_RES(trap_brkpt(), trap_brkpt_received == 1);
+	TEST_RES(trap_trace(), trap_trace_received == 1);
+}
+END_TEST()

